File: app/src/main/java/kvj/taskw/ui/EditorActivity.java
package kvj.taskw.ui;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ProgressBar;

import kvj.taskw.data.UUIDBundleAdapter;
import org.kvj.bravo7.form.BundleAdapter;
import org.kvj.bravo7.form.FormController;
import org.kvj.bravo7.form.impl.ViewFinder;
import org.kvj.bravo7.form.impl.bundle.StringBundleAdapter;
import org.kvj.bravo7.form.impl.widget.TransientAdapter;
import org.kvj.bravo7.log.Logger;
import org.kvj.bravo7.util.DataUtil;
import org.kvj.bravo7.util.Tasks;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

import kvj.taskw.App;
import kvj.taskw.R;
import kvj.taskw.data.AccountController;
import kvj.taskw.data.Controller;
import kvj.taskw.ui.AppActivity;

/**
 * Created by kvorobyev on 11/21/15.
 */
public class EditorActivity extends AppActivity {

    private Toolbar toolbar = null;
    private Editor editor = null;
    private FormController form = new FormController(new ViewFinder.ActivityViewFinder(this));
    Controller controller = App.controller();
    Logger logger = Logger.forInstance(this);
    private List<String> priorities = null;
    private AccountController.TaskListener progressListener = null;
    private AccountController ac = null;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_editor);
        toolbar = findViewById(R.id.toolbar);
        editor = (Editor) getSupportFragmentManager().findFragmentById(R.id.editor_editor);
        ProgressBar progressBar = findViewById(R.id.progress);
        setSupportActionBar(toolbar);
        logger.e("OnCreate", savedInstanceState);
        form.add(new TransientAdapter<>(new StringBundleAdapter(), null), App.KEY_ACCOUNT);
        form.add(new TransientAdapter<>(new UUIDBundleAdapter(), null), App.KEY_EDIT_UUID);
        form.add(new TransientAdapter<>(new BundleAdapter<Bundle>() {
            @Override
            public Bundle get(Bundle bundle, String name, Bundle def) {
                return bundle.getBundle(name);
            }

            @Override
            public void set(Bundle bundle, String name, Bundle value) {
                bundle.putBundle(name, value);
            }
        }, null).oneShot(), App.KEY_EDIT_DATA);
        form.add(new TransientAdapter<>(new BundleAdapter<ArrayList<String>>() {
            @Override
            public ArrayList<String> get(Bundle bundle, String name, ArrayList<String> def) {
                return bundle.getStringArrayList(name);
            }

            @Override
            public void set(Bundle bundle, String name, ArrayList<String> value) {
                bundle.putStringArrayList(name, value);
            }
        }, null).oneShot(), App.KEY_EDIT_DATA_FIELDS);
        editor.initForm(form);
        form.load(this, savedInstanceState);
        ac = controller.accountController(form);
        if (null == ac) {
            finish();
            controller.messageShort("Invalid arguments");
            return;
        }
        toolbar.setSubtitle(ac.name());
        progressListener = MainActivity.setupProgressListener(this, progressBar);
        new Tasks.ActivitySimpleTask<List<String>>(this){

            @Override
            protected List<String> doInBackground() {
                return ac.taskPriority();
            }

            @Override
            public void finish(List<String> result) {
                editor.setupPriorities(result);
                priorities = result;
                form.load(EditorActivity.this, savedInstanceState, App.KEY_EDIT_PRIORITY);
                editor.show(form);
                Bundle formData = form.getValue(App.KEY_EDIT_DATA);
                List<String> fields = form.getValue(App.KEY_EDIT_DATA_FIELDS);
                logger.d("Edit:", formData, fields);
                if (null != formData && null != fields) { // Have data
                    for (String f : fields) { // $COMMENT
                        form.setValue(f, formData.getString(f));
                    }
                }
            }
        }.exec();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        form.save(outState);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_editor, menu);
        if (null != editor && !editor.adding(form)) { // New item mode
            menu.findItem(R.id.menu_tb_add_another).setVisible(false);
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_tb_save:
                doSave(false);
                break;
            case R.id.menu_tb_add_another:
                doSave(true);
                break;
            case R.id.menu_tb_add_shortcut:
                createShortcut();
                break;
        }
        return true;
    }

    private void createShortcut() {
        Bundle bundle = new Bundle();
        form.save(bundle);
        bundle.remove(App.KEY_EDIT_UUID); // Just in case
        final Intent shortcutIntent = new Intent(this, EditorActivity.class);
        shortcutIntent.putExtras(bundle);
        shortcutIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        controller.input(this, "Shortcut name:", ac.name(), new DataUtil.Callback<CharSequence>() {

            @Override
            public boolean call(CharSequence value) {
                controller.createShortcut(shortcutIntent, value.toString().trim());
                return true;
            }
        }, null);
    }

    @Override
    protected void onResume() {
        super.onResume();
        ac.listeners().add(progressListener, true);
    }

    @Override
    protected void onPause() {
        super.onPause();
        ac.listeners().remove(progressListener);
    }

    @Override
    public void onBackPressed() {
        if (!form.changed()) { // No changes - just close
            super.onBackPressed();
            return;
        }
        logger.d("Changed:", form.changes());
        controller.question(this, "There are some changes, discard?", new Runnable() {

            @Override
            public void run() {
                EditorActivity.super.onBackPressed();
            }
        }, null);
    }
    
    private String propertyChange(String key, String modifier) {
        String value = form.getValue(key);
        if (TextUtils.isEmpty(value)) {
            value = "";
        }
        return String.format("%s:%s", modifier, value);
    }

    private String save() {
        if (!form.changed()) { // No change - no save
            return "Nothing has been changed";
        }
        String description = form.getValue(App.KEY_EDIT_DESCRIPTION);
        if (TextUtils.isEmpty(description)) { // Empty desc
            return "Description is mandatory";
        }
        List<String> changes = new ArrayList<>();
        for (String key : form.changes()) { // Make changes
            if (App.KEY_EDIT_DESCRIPTION.equals(key)) { // Direct
                changes.add(AccountController.escape(description));
            }
            if (App.KEY_EDIT_PROJECT.equals(key)) { // Direct
                changes.add(propertyChange(key, "project"));
            }
            if (App.KEY_EDIT_DUE.equals(key)) { // Direct
                changes.add(propertyChange(key, "due"));
            }
            if (App.KEY_EDIT_SCHEDULED.equals(key)) { // Direct
                changes.add(propertyChange(key, "scheduled"));
            }
            if (App.KEY_EDIT_WAIT.equals(key)) { // Direct
                changes.add(propertyChange(key, "wait"));
            }
            if (App.KEY_EDIT_UNTIL.equals(key)) { // Direct
                changes.add(propertyChange(key, "until"));
            }
            if (App.KEY_EDIT_RECUR.equals(key)) { // Direct
                changes.add(propertyChange(key, "recur"));
            }
            if (App.KEY_EDIT_PRIORITY.equals(key)) { // Direct
                changes.add(String.format("priority:%s", priorities
                    .get(form.getValue(App.KEY_EDIT_PRIORITY, Integer.class))));
            }
            if (App.KEY_EDIT_TAGS.equals(key)) { // Direct
                List<String> tags = new ArrayList<>();
                String tagsStr = form.getValue(App.KEY_EDIT_TAGS);
                Collections.addAll(tags, tagsStr.split(" "));
                changes.add(String.format("tags:%s", MainListAdapter.join(",", tags)));
            }
        }
        UUID uuid = form.getValue(App.KEY_EDIT_UUID);
        boolean completed = form.getValue(App.KEY_EDIT_STATUS, Integer.class) > 0;
        logger.d("Saving change:", uuid, changes, completed);
        if (uuid == null) { // Add new
            return completed? ac.taskLog(changes): ac.taskAdd(changes);
        } else {
            return ac.taskModify(uuid, changes);
        }
    }

    private void doSave(final boolean addAnother) {
        new Tasks.ActivitySimpleTask<String>(this) {

            @Override
            protected String doInBackground() {
                return save();
            }

            @Override
            public void finish(String result) {
                if (!TextUtils.isEmpty(result)) { // Failed
                    controller.messageLong(result);
                } else {
                    controller.messageShort("Task added");
                    EditorActivity.this.setResult(Activity.RESULT_OK);
                    if (addAnother) { // Keep everything except description
                        form.setValue(App.KEY_EDIT_DESCRIPTION, "");
                        form.getView(App.KEY_EDIT_DESCRIPTION).requestFocus();
                    } else { // Finish activity
                        EditorActivity.this.finish();
                    }
                }
            }
        }.exec();
    }

}

File: app/src/main/java/kvj/taskw/ui/EditorActivity.kt
package kvj.taskw.ui

import java.util.Calendar
import java.util.Date

import android.app.Activity
import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Intent
import android.os.Bundle
import android.text.format.DateFormat
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ArrayAdapter
import android.widget.EditText

import timber.log.Timber

import org.kvj.bravo7.util.DataUtil

import kvj.taskw.data.Task
import kvj.taskw.data.Task.Companion.Status
import kvj.taskw.App
import kvj.taskw.R
import kvj.taskw.data.AccountController

import kotlinx.android.synthetic.main.activity_editor.*
import kotlinx.android.parcel.Parcelize
import java.lang.Exception

class EditorActivity : AppForm<EditorActivity.Form>() {
    override val layout = R.layout.activity_editor

    private var priorities = listOf<String>()
    private lateinit var progressListener: AccountController.TaskListener
    private lateinit var ac: AccountController

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setSupportActionBar(toolbar)

        progressListener = MainActivity.setupProgressListener(this, progress)
        GetPrioritiesTask(this).execute()

        setupDatePicker(editor_due_btn, due)
        setupDatePicker(editor_wait_btn, wait)
        setupDatePicker(editor_scheduled_btn, scheduled)
        setupDatePicker(editor_until_btn, until)
    }

    override fun onResume() {
        super.onResume()
        ac.listeners().add(progressListener, true)
    }

    override fun onPause() {
        super.onPause()
        ac.listeners().remove(progressListener)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_editor, menu)
        menu.findItem(R.id.menu_tb_add_another).isVisible = (data.task != null)

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_tb_save ->         submit(false)
            R.id.menu_tb_add_another ->  submit(false)
            R.id.menu_tb_add_shortcut -> createShortcut()
        }

        return true
    }

    override fun loadFromForm() {
        ac = controller.accountController(data.account)
        toolbar.subtitle = ac.name()

        val task = data.task ?: return

        description.setText(task.description)
        complete.isChecked = task.status == Status.COMPLETED
        project.setText(task.project)
        tags.setText(MainListAdapter.join(",", task.tags))
        recur.setText(task.recur)

        due.setText(data.due ?: MainListAdapter.formatDate(task.due))
        wait.setText(data.wait ?: MainListAdapter.formatDate(task.wait))
        scheduled.setText(data.scheduled ?: MainListAdapter.formatDate(task.scheduled))
        until.setText(data.until ?: MainListAdapter.formatDate(task.until))

        data.original = data.original ?: task.copy()
    }

    override fun saveToForm() {
        data.task?.let {
            it.description = description.text.toString()
            it.status = if (complete.isChecked) Status.COMPLETED else Status.PENDING
            it.project = project.text.toString()
            it.tags = tags.text.toString().split("([,;]\\s*|\\s+)".toRegex())
            it.priority = priority.selectedItem as String
            it.recur = recur.text.toString()
        }

        data.let {
            it.due = due.text.toString()
            it.wait = wait.text.toString()
            it.scheduled = scheduled.text.toString()
            it.until = until.text.toString()
        }

        data.let {
            it.due = due.text.toString()
            it.wait = wait.text.toString()
            it.scheduled = scheduled.text.toString()
            it.until = until.text.toString()
        }
    }

    override fun hasChanges(): Boolean {
        return !data.due.isNullOrEmpty()
                || !data.wait.isNullOrEmpty()
                || !data.scheduled.isNullOrEmpty()
                || !data.until.isNullOrEmpty()
                || (data.task?.equals(data.original) ?: true)
    }

    private fun submit(addAnother: Boolean) {
        super.submit()
        SaveTask(this, addAnother).execute()
    }

    private fun createShortcut() {
        saveToForm()

        val shortcutIntent = Intent(this, EditorActivity::class.java).apply {
            putExtra(App.KEY_EDIT_DATA, data)
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }

        val callback = DataUtil.Callback<CharSequence> { value ->
            controller.createShortcut(shortcutIntent, value.toString().trim { it <= ' ' })
            true
        }

        controller.input(this, "Shortcut name:", ac.name(), callback, null)
    }

    private fun setupDatePicker(button: View, input: EditText) {
        fun getCalendar(input: EditText): Calendar {
            val date = try {
                MainListAdapter.formattedFormat.parse(input.text.toString().trim()) ?: Date()
            } catch (ex: Exception) {
                Date()
            }

            return Calendar.getInstance().apply { time = date }
        }

        button.setOnClickListener {
            val calendar = getCalendar(input)

            DatePickerDialog(this, { _, year, month, day ->
                calendar.set(year, month, day)
                input.setText(MainListAdapter.formattedFormat.format(calendar.time))
            }, calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH)).show()
        }

        button.setOnLongClickListener {
            val calendar = getCalendar(input)
            val format24 = DateFormat.is24HourFormat(this)

            TimePickerDialog(this, TimePickerDialog.OnTimeSetListener { _, hour, minute ->
                calendar.apply {
                    set(Calendar.HOUR_OF_DAY, hour)
                    set(Calendar.MINUTE, minute)
                    set(Calendar.SECOND, 0)
                }

                input.setText(MainListAdapter.formattedISO.format(calendar.time))
            }, calendar.get(Calendar.HOUR_OF_DAY), calendar.get(Calendar.MINUTE), format24).show()

            true
        }
    }

    fun save(): String? {
        val changes = mutableListOf<String>()

        val map = mapOf(
            "description" to Pair(data.task?.description, data.original?.description),
            "project" to Pair(data.task?.project, data.original?.project),
            "priority" to Pair(data.task?.priority, data.original?.priority),
            "due" to Pair(data.due, data.original?.due),
            "scheduled" to Pair(data.scheduled, data.original?.scheduled),
            "wait" to Pair(data.wait, data.original?.wait),
            "until" to Pair(data.until, data.original?.until),
            "recur" to Pair(data.task?.recur, data.original?.recur),
            "tags" to Pair(data.task?.tags?.joinToString(separator = ","), data.original?.tags?.joinToString(separator = ","))
        )

        for ((key, pair) in map) {
            if (pair.first == pair.second) continue

            val value = when (pair.first) {
                null -> ""
                else -> AccountController.escape(pair.first.toString())
            }

            changes.add("$key:$value")
        }

        val completed = data.task?.status == Status.COMPLETED
        Timber.d("Saving change: UUID %s, changes %s %s", data.task?.uuid, changes, completed)

        data.task?.uuid?.let { return ac.taskModify(it, changes) }
        return if (completed) ac.taskLog(changes) else ac.taskAdd(changes)
    }

    @Parcelize
    data class Form @JvmOverloads constructor(
            val account: String,
            var task: Task? = null,
            var original: Task? = null,
            var due: String? = null,
            var wait: String? = null,
            var scheduled: String? = null,
            var until: String? = null
    ) : FormData

    companion object {
        @JvmStatic
        fun start(activity: Activity, data: Form) {
            val intent = Intent(activity, EditorActivity::class.java).apply {
                putExtra(App.KEY_EDIT_DATA, data)
            }

            activity.startActivityForResult(intent, App.EDIT_REQUEST)
        }

        private class GetPrioritiesTask(activity: EditorActivity)
            : StaticAsyncTask<EditorActivity, Void, Void, List<String>>(activity) {

            override fun EditorActivity.background(vararg params: Void): List<String>
                = ac.taskPriority()

            override fun EditorActivity.finish(result: List<String>) {
                priorities = result
                priority.adapter = ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, priorities)
                priority.setSelection(priorities.indexOf(data.task?.priority ?: ""))
            }
        }

        private class SaveTask(context: EditorActivity,
                               val addAnother: Boolean)
            : StaticAsyncTask<EditorActivity, Void, Void, String?>(context) {

            override fun EditorActivity.background(vararg params: Void): String? {
                return save()
            }

            override fun EditorActivity.finish(result: String?) {
                if (!result.isNullOrBlank()) {
                    controller.messageLong(result)
                    return
                }

                controller.messageShort(
                    getString(if (data.task != null) R.string.edit_task_success else R.string.add_task_success))

                setResult(Activity.RESULT_OK)

                if (!addAnother) {
                    finish()
                    return
                }

                data.task = null
                description.requestFocus()
            }
        }
    }
}
